### PR TITLE
Add static_map::is_absent

### PR DESCRIPTION
**Pull request by vyasr**
This PR adds an operation that is the complement of `contains`.

### CODE REVIEW

File: include/cuco/static_map.cuh [hunk 1]
```diff
@@ -276,6 +276,33 @@ class static_map {
...
 
+  /**
+   * @brief Indicates whether the keys in the range `[first, last)` are missing from the map.
+   *
+   * Writes a `bool` to `(output + i)` indicating if the key `*(first + i)` does not exist in the map.
+   *
+   * @tparam InputIt Device accessible input iterator whose `value_type` is
+   * convertible to the map's `key_type`
+   * @tparam OutputIt Device accessible output iterator whose `value_type` is
+   * convertible to the map's `mapped_type`
```
Comment (jrhemstad) [inline include/cuco/static_map.cuh]:
```suggestion
   * @tparam OutputIt Device accessible output iterator whose `value_type` is
   * constructible from `bool`. 
```

Comment (jrhemstad) [reply include/cuco/static_map.cuh]: I see this is a problem with `contains` above too.

File: include/cuco/static_map.cuh [hunk 2]
```diff
@@ -276,6 +276,33 @@ class static_map {
...
+   * @tparam Hash Unary callable type
+   * @tparam KeyEqual Binary callable type
+   * @param first Beginning of the sequence of keys
+   * @param last End of the sequence of keys
+   * @param output_begin Beginning of the sequence of booleans for the presence of each key
```
Comment (jrhemstad) [inline include/cuco/static_map.cuh]: Strictly speaking, the output sequence doesn't need to literally be bools, it just needs to be assignable from bools. 
```suggestion
   * @param output_begin Beginning of the output sequence to which a `bool` is assigned indicating the presence of each key 
```

File: include/cuco/detail/static_map.inl [hunk 3]
```diff
@@ -109,11 +109,30 @@ void static_map<Key, Value, Scope, Allocator>::contains(
...
+template <typename InputIt, typename OutputIt, typename Hash, typename KeyEqual>
+void static_map<Key, Value, Scope, Allocator>::is_absent(
+  InputIt first, InputIt last, OutputIt output_begin, Hash hash, KeyEqual key_equal, cudaStream_t stream)
+{
+  auto num_keys = std::distance(first, last);
+  if (num_keys == 0) { return; }
+
+  auto const block_size = 128;
+  auto const stride     = 1;
+  auto const tile_size  = 4;
```
Comment (PointKernel) [inline include/cuco/detail/static_map.inl:127]:
```suggestion
  auto constexpr block_size = 128;
  auto constexpr stride     = 1;
  auto constexpr tile_size  = 4;
```
The same for bulk `contains`

File: include/cuco/detail/static_map.inl [hunk 4]
```diff
@@ -109,11 +109,30 @@ void static_map<Key, Value, Scope, Allocator>::contains(
...
+  auto const grid_size  = (tile_size * num_keys + stride * block_size - 1) / (stride * block_size);
+  auto view             = get_device_view();
+
+  detail::check_contains<block_size, tile_size, false>
```
Comment (PointKernel) [inline include/cuco/detail/static_map.inl:131]:
```suggestion
  auto constexpr check_presense = false;
  detail::check_contains<block_size, tile_size, check_presense>
```
Avoid raw boolean value for readability

File: include/cuco/detail/static_map_kernels.cuh [hunk 5]
```diff
@@ -278,11 +278,12 @@ __global__ void find(
...
+ * @brief Indicates whether the keys in the range `[first, last)` are contained in (or missing from) the map.
  *
- * Writes a `bool` to `(output + i)` indicating if the key `*(first + i)` exists in the map.
+ * Writes a `bool` to `(output + i)` indicating if the key `*(first + i)` exists/does not exist in the map.
  *
  * @tparam block_size The size of the thread block
+ * @tparam present If false, check whether elements are absent rather than checking for their presence.
  * @tparam InputIt Device accessible input iterator whose `value_type` is
  * convertible to the map's `key_type`
  * @tparam OutputIt Device accessible output iterator whose `value_type` is
```
Comment (PointKernel) [inline include/cuco/detail/static_map_kernels.cuh:289]:
```cpp
* @tparam OutputIt Device accessible output iterator whose `value_type` is
* convertible from `bool`
```

File: include/cuco/detail/static_map.inl [hunk 6]
```diff
@@ -109,11 +109,30 @@ void static_map<Key, Value, Scope, Allocator>::contains(
   auto const grid_size  = (tile_size * num_keys + stride * block_size - 1) / (stride * block_size);
   auto view             = get_device_view();
 
-  detail::contains<block_size, tile_size>
+  detail::check_contains<block_size, tile_size, true>
```
Comment (PointKernel) [inline include/cuco/detail/static_map.inl:112]:
```suggestion
  auto constexpr check_presense = true;
  detail::check_contains<block_size, tile_size, check_presense>
```

Comment (jrhemstad) [reply include/cuco/detail/static_map.inl:112]: or just use an `enum class`
```
enum class find{
  PRESENT, ABSENT
};
```

File: include/cuco/detail/static_map_kernels.cuh [hunk 7]
```diff
@@ -354,12 +361,13 @@ __global__ void contains(
...
 template <std::size_t block_size,
           uint32_t tile_size,
+          bool present,
           typename InputIt,
           typename OutputIt,
           typename viewT,
           typename Hash,
           typename KeyEqual>
-__global__ void contains(
+__global__ void check_contains(
```
Comment (jrhemstad) [inline include/cuco/detail/static_map_kernels.cuh:370]: This sounds a bit more natural to me. 
```suggestion
__global__ void check_presence(
```

Comment (jrhemstad) [reply include/cuco/detail/static_map_kernels.cuh:370]: or `check_existence`

File: include/cuco/detail/static_map.inl (hunk 6 above)
Comment (jrhemstad) [reply include/cuco/detail/static_map.inl:112]: or:
```
enum class exists{ YES, NO };
enum class present{ YES, NO};
```

**Comment by vyasr** [comment]
@jrhemstad @PointKernel is this feature something of interest? We found a different solution for the use case in cudf IIRC, unclear if this is worth tackling again or if it should be closed for now.

**Comment by PointKernel** [comment]
I'm closing this PR for now since `cuco::static_map` will be completely rewritten. 
We could reevaluate the use of this functionality during refactoring.

**Pull request closed without merging**